### ISSUE TITLE
Impulse UGEN sets the presample to the wrong value when frequency equals zero

### ISSUE THREAD

**Issue by cianoc** [supercollider/supercollider#4075]
Environment
-----------

* SuperCollider 3.93
* OSX 10.13.6

Steps to reproduce (for bugs)
-----------------------------

```supercollider
(
{
  var sig1 = EnvGen.kr(Env.perc(0.01, 0.04), 1) * SinOsc.ar(440);
  var sig2 = EnvGen.kr(Env.perc(0.01, 0.04), Impulse.kr(0)) * SinOsc.ar(440);
  [sig1,sig2]
}.plot)
```

**Graph:**
![screen shot 2018-09-21 at 7 54 12 am](https://user-images.githubusercontent.com/355604/45879761-aca23380-bd73-11e8-9efb-dc6e15cfdc2c.png)

------------------------

Expected Behavior
-----------------
The two audio signals should be identical.

Current Behavior
----------------
Currently the envelope triggered by Impulse.kr is a control period later.

**Comment by cianoc** [comment]
This is caused by the fact that for ```Impulse.kr``` the pre-sample values are *always* set to 0. An easy solution to this would be simply to change the pre-sample value to 1 when frequency is 0, and also set the phase appropriately.

*However* this would change the existing behavior of ```Impulse.kr``` for existing synthdefs. My personal opinion is that this is acceptable as the behavior above is clearly wrong, is confusing (particularly as Impulse.kr fires an impulse at 0 seconds), inconsistent, un-mathematical, etc. I understand backwards compatibility concerns, but I don't think preserving clearly wrong behavior is the way to go. Obviously others may disagree here.

I know *how* to fix this, and if there is some consensus that this should be fixed then I'll do so. But I don't want to waste time on this if people want to preserve the existing behavior.

**Comment by cianoc** [comment]
Sorry I should have been clearer. The only change of the existing behavior would be that it would fix the bug. However this bug has been around presumably for ever, so it could change how existing pieces sound if they rely upon this behavior. This feels like a small risk to me, but I know backwards compatibility is important.

So to clarify - my proposed fix wouldn't affect anything except when Impulse is passed a frequency of 0. Currently the envelope only begins on the second control block (e.g. sample 65) - which is clearly wrong. I propose changing this so that the envelope would begin on the first control block (e.g. sample 1) - which is expected behavior.

**Comment by telephon** [comment]
>So to clarify - my proposed fix wouldn't affect anything except when Impulse is passed a frequency of 0. 

Yes, definitely. I was really just trying to imagine how you could rely on the bug. Any idea?

Btw. I am the one who implemented the `Impulse.ar(0)`-feature (and this bug slipped in because the whole initialization was still incoherently treated all over the place).

**Comment by cianoc** [comment]
> Yes, definitely. I was really just trying to imagine how you could rely on the bug. Any idea?

Not really. I'm of the opinion that it should be fixed. I was just acknowledging the fact that in certain cases it *might* affect the sound of existing pieces slightly. But it's hard for to believe that anyone composed a piece knowing about this bug and deliberately exploiting it.

**Comment by cianoc** [comment]
Okay, this seems to relate to: [Issue #2343](https://github.com/supercollider/supercollider/issues/2343). That discussion kind of fizzled out without any resolution. I agree with James the current situation with Impulse is ludicrous, and it makes certain things impossible. But there doesn't seem to be any resolution on what the best way to proceed here is.

Myself I liked James' solution of just fixing Impulse.

**Comment by cianoc** [comment]
Okay, this seems to relate to: [Issue #2343](https://github.com/supercollider/supercollider/issues/2343). That discussion kind of fizzled out without any resolution. I agree with James the current situation with Impulse is ludicrous, and it makes certain things impossible. But there doesn't seem to be any resolution on what the best way to proceed here is.

Myself I liked James' solution of just fixing Impulse.

**Comment by telephon** [comment]
It is a complicated discussion and I am not sure I can make up my mind easily.  Closest to a conclusion seemed:

>The simplest solution is that the UGen expecting triggers inits prevTrig to 0 in the Ctor.

If I understand correctly, the problem *should* be that `EnvGen` doesn't initialize its previous gate value to zero? If you could check this, that would be great. If it turns out not to be the case, then we have to rethink the trigger semantics.

**Comment by telephon** [comment]
Agreed. Let's fix the impulse implementation first.

**Comment by mtmccrea** [comment]
I wanted to raise a cautionary point about using `UGen_next_x(unit, 1)` to initialize the state of a UGen, and please elaborate if I'm missing something...

Many `next_x` calc functions are setting member variables of the unit: `unit->mPhase = phase;` in the case of `Impulse_next_k`. During the next sample's calculation, this state is expected to have already "happened", i.e. it's now a past state. However, the output of `UGen_next_x(unit, 1)` isn't actually ever used—it's recalculated at the first sample of the the first control full control loop. So the unit's state was moved forward in time `t = 1` (`mPhase` in this case), but once recalculated for true output, the resulting sample represents `t = 0`;

Please correct me if I'm mistaken, as this has had my brain in knots after reading through some very looong threads

**NOTE** this also has implications about how `SCUnit`'s (re-)definition of `set_calc_function` automatically calls `(mCalcFunc)(this, 1);`, which may not be best practice, giving all the special care needed for this initial sample calculation!

**Comment by mtmccrea** [comment]
Also, I think `ToggleFF` may similarly require whatever solution is agreed on for `Impulse`.

**Comment by mtmccrea** [comment]
I wanted to mention here that I've been working on a rewrite of `Impulse` and I am progressing, though there are many use cases to account for, especially considering that the phase offset is now modulatable.

It's proven to be a bit tricky, and at this point I'll need to get some input from some other folks on the dev list, but I'll update here if I resolve it.

**Comment by mtmccrea** [comment]
>  I'll update here if I resolve it.

Update: first step in resolving this issue has been taken - #4150

**Comment by mtmccrea** [comment]
This behavior is explained by the fact that the envelope is k-rate.

If you plot the envelopes alone, you see they behave as expected:
```supercollider
(
{
  var sig1 = EnvGen.kr(Env.perc(0.01, 0.04), 1);
  var sig2 = EnvGen.kr(Env.perc(0.01, 0.04), Impulse.kr(0));
  [sig1,sig2]
}.plot(0.05).plotMode_(\plines)
)
```

<img width="512" alt="Image" src="https://github.com/user-attachments/assets/2f1dfcc0-ac9e-416c-8335-cd62c637351c" />

And they both match, as they should (so Impulse and the EnvGen trigger behave properly).

Consider though what the envelope actually looks like when applied to an audio-rate signal:
```supercollider
(
{
  var sig = EnvGen.kr(Env.perc(0.01, 0.04), Impulse.kr(0));
  [sig]
}.plot(0.02).plotMode_(\levels);

{
  var sig = EnvGen.kr(Env.perc(0.01, 0.04), Impulse.kr(0)) * SinOsc.ar(440);
  sig
}.plot(0.02).plotMode_(\plines);
)
```
<img width="752" alt="Image" src="https://github.com/user-attachments/assets/4d38a14d-866a-45ed-88b6-5745f8b31ce7" />

k-rate signals are sample-and-hold unless they're send through `K2A.ar`, though in that case the signal will be effectively "delayed" in the course of interpolating to subsequent points.

So, unless I'm overlooking something, I think this issue can be closed.

**Comment by mtmccrea** [comment]
I believe the "delay" comes from the fact that the envelope, which is k-rate, is upsampled to a-rate via `K2A`, which interpolates from the previous to the next value over one control period. The first block is interpolating from the initial value to the first value, which is the same (0 to 0).

**Comment by elgiano** [comment]
Maybe I was not clear, try this example:
```supercollider
{
  var env = Env.perc(1e-4,1e-2);
  [EnvGen.ar(env), EnvGen.kr(env)]
}.plot
```

![Image](https://github.com/user-attachments/assets/de5b6dbc-122d-407a-90b9-e08840009431)

for the same envelope, the ar version starts immediately, while the kr version starts late. I can't find the K2A you're referring to anywhere in the chain, could you point me out to it?

**Comment by mtmccrea** [comment]
> {
>   var env = Env.perc(1e-4,1e-2);
>   [EnvGen.ar(env), EnvGen.kr(env)]
> }.plot
> 
> for the same envelope, the ar version starts immediately, while the kr version starts late. I can't find the K2A you're referring to anywhere in the chain, could you point me out to it?

Because you are plotting both k-rate and a-rate signals together, it upsamples the k-rate signals using K2A, which imposes a 1-block lag on the values.

**Comment by elgiano** [comment]
Ah wow, perfect, I didn't think of checking Plot! I think we can close this issue then (and open one about documenting plot and K2A?)
